lib/api/wasm_c.cpp

```diff
@@ -518,6 +537,684 @@ WASM_DECLARE_C_GET_ATTR(exporttype, wasm_externtype_t *, type, static, nullptr,
+    return static_cast<const wasm_##name##_t *>(                               \
+        static_cast<const Name *>(static_cast<const wasm::Ref *>(r)));         \
+  }
+
+#define WASM_DEFINE_SHARABLE_REF(name, Name)                                   \
+  WASM_DEFINE_REF(name, Name)                                                  \
+  WASM_DECLARE_OWN_IMPL(shared_##name, wasm::Shared<wasm::Module>)
+
+WASM_DEFINE_REF_BASE(ref, wasm::Ref)
+
+// Values
+
+extern "C++" {
+
+inline bool is_empty(wasm_val_t v) {
+  return !is_ref(static_cast<wasm::ValKind>(v.kind)) || !v.of.ref;
+}
+// TODO: Find solution to:
+// missing initializer for member ‘wasm_val_t::of’
+// [-Werror=missing-field-initializers]
+
+inline wasm_val_t hide_val(wasm::Val v) {
+  //  wasm_val_t v2 = {static_cast<wasm_valkind_t>(v.kind())};
+  wasm_valkind_t kind = static_cast<wasm_valkind_t>(v.kind());
+  switch (v.kind()) {
+  case wasm::ValKind::I32:
+    return {kind, {.i32 = v.i32()}};
+  case wasm::ValKind::I64:
+    return {kind, {.i64 = v.i64()}};
+  case wasm::ValKind::F32:
+    return {kind, {.f32 = v.f32()}};
+  case wasm::ValKind::F64:
+    return {kind, {.f64 = v.f64()}};
+  case wasm::ValKind::ANYREF:
+  case wasm::ValKind::FUNCREF:
+    return {kind, {.ref = static_cast<wasm_ref_t *>(v.ref())}};
+  default:
+    break;
+    assert(false);
+  }
+  return {kind, {.i32 = v.i32()}};
+}
+
+inline wasm_val_t release_val(wasm::Val v) {
+  //  wasm_val_t v2 = {static_cast<wasm_valkind_t>(v.kind()), {v.i32()}};
+  wasm_valkind_t kind = static_cast<wasm_valkind_t>(v.kind());
+  switch (v.kind()) {
+  case wasm::ValKind::I32:
+    return {kind, {.i32 = v.i32()}};
+  case wasm::ValKind::I64:
+    return {kind, {.i64 = v.i64()}};
+  case wasm::ValKind::F32:
+    return {kind, {.f32 = v.f32()}};
+  case wasm::ValKind::F64:
+    return {kind, {.f64 = v.f64()}};
+  case wasm::ValKind::ANYREF:
+  case wasm::ValKind::FUNCREF:
+    return {kind,
+            {.ref = static_cast<wasm_ref_t *>(v.release_ref().release())}};
+  default:
+    assert(false);
+  }
+  return {kind, {.i32 = v.i32()}};
+}
+
+inline wasm::Val adopt_val(wasm_val_t v) {
+  switch (static_cast<wasm::ValKind>(v.kind)) {
+  case wasm::ValKind::I32:
+    return wasm::Val(v.of.i32);
+  case wasm::ValKind::I64:
+    return wasm::Val(v.of.i64);
+  case wasm::ValKind::F32:
+    return wasm::Val(v.of.f32);
+  case wasm::ValKind::F64:
+    return wasm::Val(v.of.f64);
+  case wasm::ValKind::ANYREF:
+  case wasm::ValKind::FUNCREF:
+    return wasm::Val(wasm::make_own(v.of.ref));
+  default:
+    assert(false);
+  }
+  return wasm::Val(v.of.i32);
+}
+
+struct borrowed_val {
+  wasm::Val it;
+
+  borrowed_val(wasm::Val &&v) : it(std::move(v)) {}
+
+  borrowed_val(borrowed_val &&that) : it(std::move(that.it)) {}
+
+  ~borrowed_val() {
+    if (it.is_ref())
+      it.release_ref().release();
+  }
+};
+
+inline borrowed_val borrow_val(const wasm_val_t *v) {
```

Comment:
Merge this function into the API function body.

Reply:
Does this function have to be merged
even if it is used more than one time?

Reply:
Move into the anonymous namespace on the top.